### ISSUE TITLE
GenericTypeReflector.transform visitor visitVariable cannot change type away from variable type

### ISSUE THREAD

**Issue by stevenschlansker** [leangen/geantyref#13]
Hi @kaqqao thank you for your continued help, I wonder what you think of this.

Given the class,
```java
public class GenericBean<T extends String> {
    public List<T> getProperty() {
        return Collections.emptyList();
    }
}
```
I want to figure out the most specific type for the return value of `getProperty` given that I don't have the parameterized `GenericBean<T>`, only `GenericBean.class` - replacing `T` with `String` (the bound) and then figure out `List<String>` as the property type.

To do this, I am trying to use `GenericTypeReflector.transform` and override `visitVariable` to replace variables with their bound.

```java
private AnnotatedType resolveBounds(final Type type) {
    return GenericTypeReflector.transform(GenericTypeReflector.annotate(type), new TypeVisitor() {
        @Override
        protected AnnotatedType visitVariable(final AnnotatedTypeVariable type) {
            return type.getAnnotatedBounds()[0]; // T extends String -> String
        }
    });
}

        assertThat(Arrays.stream(GenericBean.class.getTypeParameters())
                .map(this::resolveBounds)
                .toArray(AnnotatedType[]::new))
            .containsExactly(GenericTypeReflector.annotate(String.class));
```

Unfortunately, despite this being statically safe, it does not actually work:

```
java.lang.ClassCastException: class sun.reflect.annotation.AnnotatedTypeFactory$AnnotatedTypeBaseImpl cannot be cast to class java.lang.reflect.AnnotatedTypeVariable (sun.reflect.annotation.AnnotatedTypeFactory$AnnotatedTypeBaseImpl and java.lang.reflect.AnnotatedTypeVariable are in module java.base of loader 'bootstrap')
	at io.leangen.geantyref.TypeVisitor.visitCaptureType(TypeVisitor.java:75)
	at io.leangen.geantyref.GenericTypeReflector.transform(GenericTypeReflector.java:1129)
	at io.leangen.geantyref.TypeVisitor.lambda$visitParameterizedType$0(TypeVisitor.java:27)
...
	at io.leangen.geantyref.TypeVisitor.visitParameterizedType(TypeVisitor.java:28)
	at io.leangen.geantyref.GenericTypeReflector.transform(GenericTypeReflector.java:1117)
	at org.jdbi.v3.core.argument.TestBeanArguments.resolveBounds(TestBeanArguments.java:292)
	at org.jdbi.v3.core.argument.TestBeanArguments.wut(TestBeanArguments.java:284)
```

It looks like the visitor assumes it can cast the result: 
```java
AnnotatedCaptureType annotatedCapture = new AnnotatedCaptureTypeImpl((CaptureType) type.getType(),
				(AnnotatedWildcardType) transform(type.getAnnotatedWildcardType(), this),
				(AnnotatedTypeVariable) transform(type.getAnnotatedTypeVariable(), this),
				null, type.getAnnotations());
```

Please let me know if I'm on the wrong track, or if there's something I'm missing.  Thank you!

**Comment by kaqqao** [comment]
I tested this out quickly on Java 8 and it seems to work as you'd expect 😓 What JDK are you on?
I fear we're getting into an ugly territory here 😓

**Comment by kaqqao** [comment]
But then again, the error doesn't seem  to suggest anything Java version specific... Maybe I misunderstood you. I'll investigate more.

**Comment by kaqqao** [comment]
Oh, I see the problem now. I deleted the previous 2 messages I wrote, as I they were just me being confused, nothing useful.
But I think the provided test case doesn't trigger it, there's got to be a wildcard somewhere to capture. But no matter, I understand the issue... I'll see what I can do.

**Comment by kaqqao** [comment]
The default `TypeVisitor` was trying to be too smart... I simplified its behavior in regards to capture types.
Additionally, I implemented a method doing exactly (I think) what you need: [`GenericTypeReflector#reduceBounded`](https://github.com/leangen/geantyref/blob/geantyref-v1.3.13/src/main/java/io/leangen/geantyref/GenericTypeReflector.java#L1153). It reduces all bounded types (variables, wildcards & captures) to their first bound. I actually want to use this this exact same thing in GraphQL SPQR instead of the current (flaky) custom implementation not relying on `transform` 😅
Released v1.3.13 with this.

**Comment by stevenschlansker** [comment]
Thank you so much!  The `reduceBounded` method seems to do exactly what I need. Jdbi now has much better type handling :smile_cat:
